modal/file_io.py

```diff
@@ -0,0 +1,335 @@
+            api_pb2.ContainerFilesystemExecRequest(
+                file_open_request=api_pb2.ContainerFileOpenRequest(path=path, mode=mode),
+                task_id=self._task_id,
+            )
+        )
+        if not resp.HasField("file_descriptor"):
+            raise FilesystemExecutionError("Failed to open file")
+        self._file_descriptor = resp.file_descriptor
+        await self._wait(resp.exec_id)
+
+    @classmethod
+    async def create(cls, path: str, mode: str, client: _Client, task_id: str) -> "_FileIO":
+        self = cls.__new__(cls)
+        self._validate_mode(mode)
+        self._client = client
+        self._task_id = task_id
+        await self._open_file(path, mode)
+        self._closed = False
+        return self
+
+    async def read(self, n: Union[int, None] = None) -> Union[bytes, str]:
+        """Read n bytes from the current position, or the entire remaining file if n is None."""
+        self._check_closed()
+        self._check_readable()
+        if n is not None and n > READ_FILE_SIZE_LIMIT:
+            raise ValueError("Read request payload exceeds 100MB limit")
+        resp = await self._client.stub.ContainerFilesystemExec(
+            api_pb2.ContainerFilesystemExecRequest(
+                file_read_request=api_pb2.ContainerFileReadRequest(file_descriptor=self._file_descriptor, n=n),
+                task_id=self._task_id,
+            )
+        )
+        return await self._wait(resp.exec_id)
+
+    async def readline(self) -> Union[bytes, str]:
+        """Read a single line from the current position."""
+        self._check_closed()
+        self._check_readable()
+        resp = await self._client.stub.ContainerFilesystemExec(
+            api_pb2.ContainerFilesystemExecRequest(
+                file_read_line_request=api_pb2.ContainerFileReadLineRequest(file_descriptor=self._file_descriptor),
+                task_id=self._task_id,
+            )
+        )
+        return await self._wait(resp.exec_id)
+
+    async def readlines(self) -> Union[List[bytes], List[str]]:
+        """Read all lines from the current position."""
+        self._check_closed()
+        self._check_readable()
+        data = await self.read()
+        if self._binary:
+            lines = cast(bytes, data).split(b"\n")
+            return [line + b"\n" for line in lines[:-1]] + ([lines[-1]] if lines[-1] else [])
+        else:
+            lines = cast(str, data).split("\n")
+            return [line + "\n" for line in lines[:-1]] + ([lines[-1]] if lines[-1] else [])
+
+    async def write(self, data: Union[bytes, str]) -> None:
+        """Write data to the current position.
+
+        Writes may not appear until the entire buffer is flushed, which
+        can be done manually with `flush()` or automatically when the file is
+        closed.
+        """
+        self._check_closed()
+        self._check_writable()
+        self._validate_type(data)
+        if isinstance(data, str):
```

Comment:
Is this the behavior of `write` in Python? If it requires you to pass bytes, I think we should too.

Reply:
`self._validate_type` enforces that the correct type is passed in, but we always convert to `bytes` when sending the rpc.

Reply:
Yeah my question is whether python `write` accept string if you're in non-bytes mode

Reply:
python `write` enforces that the correct type is passed in